src/Behat/Config/ProfileInterface.php

```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Behat\Config;
+
+interface ProfileInterface extends ConfigInterface
```

Comment:
I think we should call all of these `XXXConfigInterface` (like you have `SuiteConfigInterface` in #1540). Otherwise for things like Suite, Extension, etc there is potential for confusion - at least if just glancing at the code - about whether you're expected to pass an actual instance of the extension rather than just a config object.

Equally, I am wondering if we definitely need separate interfaces and final classes for all of these, if the interface  for most of them is only going to have `name()` and `toArray()`. If the sugar methods are only defined / implemented on the final class then it will be harder for people to find those, and the interfaces add extra noise.

Since you have to use the classes we provide to get the sugar methods, and you always have the option to pass an array (or a class of your own that doesn't satisfy any interface) into the `$settings` of the container object, there are still plenty of ways you can customise the config without needing to be able to implement an interface?

The exception to this might be for extensions and formatters, where by definition we're not in control of the available config options so interfaces would be useful to allow extensions to define a custom config object of their own.

Reply:
@acoulton For the `XXXConfigInterface` part, I'm mitigated on that. This was an error on #1540.
If we use this suffix for the interface, we need a Config suffix on implementations SuiteConfig, ProfileConfig...

But maybe that's better... To be honest, naming is never easy :+1:

Reply:
For the final classes, they contains default helpers/sugar methods, and users can define its own ones. But maybe that's too much control.

Reply:
> If we use this suffix for the interface, we need a Config suffix on implementations SuiteConfig, ProfileConfig...

I don't think we automatically *have to*. IMO they have different usecases:

* The interface is for when you're looking at the method you're calling to see what it accepts. Here, the priority is to make very clear to the end-user that `withExtension` takes an object with extension configuration, not e.g. an instance of an extension (which you might assume, if new).
* The concrete class is generally for when you're looking at / working in a config file. Here, it is already clear that you have config objects, and probably the priority is to keep the naming as concise as possible so that users are mostly looking at the actual config, rather than the boilerplate PHP structure.

So if we have interfaces, personally I would have `SuiteConfigInterface`, `ProfileConfigInterface`, `ExtensionConfigInterface` and then `SuiteConfig`, `ProfileConfig`, `ExtensionConfig` - or even just `Suite`, `Profile`, `Extension` (leaving the Config in the namespace).

Reply:
> For the final classes, they contains default helpers/sugar methods, and users can define its own ones. But maybe that's too much control.

I think it adds complexity (and a lot of very similar interfaces) for not much benefit. These objects (especially the bits that are defined by the interface) are just DTOs, and I wouldn't usually implement an interface for that.

In reality, if users want to use any of our helper methods, they'll mostly have to use our DTOs, because it's only the DTOs that have the helpers. If they don't, they can just do something like:

```
return new Behat\Config\Config([
  'default' => (new MyCustomProfileConfigObject(...))->toArray()
])
```

On the other hand, for the more common case of a user wanting to use the default helpers, it would be much easier to see that `->withProfile` takes a `Profile` or `ProfileConfig` object, and click directly through to that in their IDE rather than clicking first to an interface and then see what implements it.

> But if we define interfaces for extensions, I think it would be better to have the same control on Profiles, Suites etc.

Extensions and formatters are special cases, because unlike Profiles, Suites, etc we have no knowledge or control of what config schema they take. So it would be very common that an extension author would want to provide a config object that you could pass to `->withExtension()`.

Reply:
I removed the interface.